File: server/apps/assetmanager/schemas/holding_schemas/holding_performance_schemas.py
"""
HoldingPerformance Schemas

Pydantic schemas for the HoldingPerformance model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type

# ==========================================
# HoldingPerformance Schema (Full Representation)
# ==========================================

class HoldingPerformance(BaseModel):
    """HoldingPerformance schema - full representation"""
    id: int
    entity_id: int = Field(description="Associated entity ID")
    funding_round_id: int | None = Field(None, description="Associated funding round ID")
    report_date: date_type = Field(description="Report date")

    # Performance Metrics
    total_invested_amount: float | None = Field(None, description="Total invested amount")
    fair_value: float | None = Field(None, description="Fair value")
    cash_realized: float | None = Field(None, description="Cash realized")
    tvpi: float | None = Field(None, description="Total value to paid-in")
    dpi: float | None = Field(None, description="Distributions to paid-in")
    rvpi: float | None = Field(None, description="Residual value to paid-in")
    irr: float | None = Field(None, description="Internal rate of return")
    multiple_to_cost: float | None = Field(None, description="Multiple to cost")

    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class HoldingPerformanceCreate(BaseModel):
    """Schema for creating a new holding performance record"""
    # Required fields
    entity_id: int = Field(description="Associated entity ID")

    # Optional fields
    funding_round_id: int | None = Field(None, description="Associated funding round ID")
    report_date: date_type | None = Field(None, description="Report date")

    # Performance Metrics
    total_invested_amount: float | None = Field(None, description="Total invested amount")
    fair_value: float | None = Field(None, description="Fair value")
    cash_realized: float | None = Field(None, description="Cash realized")
    tvpi: float | None = Field(None, description="Total value to paid-in")
    dpi: float | None = Field(None, description="Distributions to paid-in")
    rvpi: float | None = Field(None, description="Residual value to paid-in")
    irr: float | None = Field(None, description="Internal rate of return")
    multiple_to_cost: float | None = Field(None, description="Multiple to cost")

class HoldingPerformanceUpdate(BaseModel):
    """Schema for updating a holding performance record"""
    entity_id: int | None = None
    funding_round_id: int | None = None
    report_date: date_type | None = None

    # Performance Metrics
    total_invested_amount: float | None = None
    fair_value: float | None = None
    cash_realized: float | None = None
    tvpi: float | None = None
    dpi: float | None = None
    rvpi: float | None = None
    irr: float | None = None
    multiple_to_cost: float | None = None

# ==========================================
# Response Types
# ==========================================

class HoldingPerformanceResponse(BaseModel):
    """Response containing a single holding performance record"""
    success: bool
    data: HoldingPerformance | None = None
    error: str | None = None

class HoldingPerformancesResponse(BaseModel):
    """Response containing multiple holding performance records"""
    success: bool
    data: list[HoldingPerformance] | None = None
    error: str | None = None

File: client/src/modules/assetmanager/schemas/holding/holding-performance.schemas.ts
/**
 * Holding Performance Schemas
 *
 * Zod validation schemas for HoldingPerformance model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/holding_models.py
 * - Schema: /server/apps/assetmanager/schemas/holding_schemas/holding_performance_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/holding_subrouters/holding_performance_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// HoldingPerformance Schema (Full Representation)
// ==========================================

/**
 * HoldingPerformance schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class HoldingPerformance(BaseModel)
 */
export const HoldingPerformanceSchema = z.object({
  id: z.number(),
  entity_id: z.number(),
  funding_round_id: z.number().nullable(),
  report_date: z.string(),

  // Performance Metrics
  total_invested_amount: z.number().nullable(),
  fair_value: z.number().nullable(),
  cash_realized: z.number().nullable(),
  tvpi: z.number().nullable(),
  dpi: z.number().nullable(),
  rvpi: z.number().nullable(),
  irr: z.number().nullable(),
  multiple_to_cost: z.number().nullable(),

  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(), // ISO datetime string from backend
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new holding performance record (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class HoldingPerformanceCreate(BaseModel)
 */
export const CreateHoldingPerformanceSchema = z.object({
  entity_id: z.number(),
  funding_round_id: z.number().nullable().optional(),
  report_date: z.string().nullable().optional(),

  // Performance Metrics
  total_invested_amount: z.number().nullable().optional(),
  fair_value: z.number().nullable().optional(),
  cash_realized: z.number().nullable().optional(),
  tvpi: z.number().nullable().optional(),
  dpi: z.number().nullable().optional(),
  rvpi: z.number().nullable().optional(),
  irr: z.number().nullable().optional(),
  multiple_to_cost: z.number().nullable().optional(),
});

/**
 * Schema for updating a holding performance record (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class HoldingPerformanceUpdate(BaseModel)
 */
export const UpdateHoldingPerformanceSchema = z.object({
  entity_id: z.number().nullable().optional(),
  funding_round_id: z.number().nullable().optional(),
  report_date: z.string().nullable().optional(),

  // Performance Metrics
  total_invested_amount: z.number().nullable().optional(),
  fair_value: z.number().nullable().optional(),
  cash_realized: z.number().nullable().optional(),
  tvpi: z.number().nullable().optional(),
  dpi: z.number().nullable().optional(),
  rvpi: z.number().nullable().optional(),
  irr: z.number().nullable().optional(),
  multiple_to_cost: z.number().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type HoldingPerformance = z.infer<typeof HoldingPerformanceSchema>;
export type CreateHoldingPerformance = z.infer<typeof CreateHoldingPerformanceSchema>;
export type UpdateHoldingPerformance = z.infer<typeof UpdateHoldingPerformanceSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single holding performance record
 * Backend equivalent: class HoldingPerformanceResponse(BaseModel)
 */
export type HoldingPerformanceResponse = {
  success: boolean;
  data?: HoldingPerformance;
  error?: string;
};

/**
 * Response containing multiple holding performance records
 * Backend equivalent: class HoldingPerformancesResponse(BaseModel)
 */
export type HoldingPerformancesResponse = {
  success: boolean;
  data?: HoldingPerformance[];
  error?: string;
};
